service/worker/scanner/shardscanner/activities.go

```diff
@@ -0,0 +1,336 @@
+	}
+	queryArgsBytes, err := json.Marshal(queryArgs)
+	if err != nil {
+		return nil, cadence.NewCustomError(ErrSerialization)
+	}
+	queryResp, err := client.QueryWorkflow(activityCtx, &shared.QueryWorkflowRequest{
+		Domain: c.StringPtr(c.SystemLocalDomainName),
+		Execution: &shared.WorkflowExecution{
+			WorkflowId: c.StringPtr(params.ScannerWorkflowWorkflowID),
+			RunId:      c.StringPtr(params.ScannerWorkflowRunID),
+		},
+		Query: &shared.WorkflowQuery{
+			QueryType: c.StringPtr(ShardCorruptKeysQuery),
+			QueryArgs: queryArgsBytes,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	queryResult := &ShardCorruptKeysQueryResult{}
+	if err := json.Unmarshal(queryResp.QueryResult, &queryResult); err != nil {
+		return nil, cadence.NewCustomError(ErrSerialization)
+	}
+	var corrupted []CorruptedKeysEntry
+	var minShardID *int
+	var maxShardID *int
+	for sid, keys := range queryResult.Result {
+		if minShardID == nil || *minShardID > sid {
+			minShardID = c.IntPtr(sid)
+		}
+		if maxShardID == nil || *maxShardID < sid {
+			maxShardID = c.IntPtr(sid)
+		}
+		corrupted = append(corrupted, CorruptedKeysEntry{
+			ShardID:       sid,
+			CorruptedKeys: keys,
+		})
+	}
+	return &FixerCorruptedKeysActivityResult{
+		CorruptedKeys:             corrupted,
+		MinShard:                  minShardID,
+		MaxShard:                  maxShardID,
+		ShardQueryPaginationToken: queryResult.ShardQueryPaginationToken,
+	}, nil
+}
+
+// FixShardActivity will fix a collection of shards.
+func FixShardActivity(
+	activityCtx context.Context,
+	params FixShardActivityParams,
+) ([]FixReport, error) {
+	heartbeatDetails := FixShardHeartbeatDetails{
+		LastShardIndexHandled: -1,
+		Reports:               nil,
+	}
+	if activity.HasHeartbeatDetails(activityCtx) {
+		if err := activity.GetHeartbeatDetails(activityCtx, &heartbeatDetails); err != nil {
+			return nil, err
+		}
+	}
+	for i := heartbeatDetails.LastShardIndexHandled + 1; i < len(params.CorruptedKeysEntries); i++ {
+		currentShardID := params.CorruptedKeysEntries[i].ShardID
+		currentKeys := params.CorruptedKeysEntries[i].CorruptedKeys
+		shardReport, err := fixShard(activityCtx, params, currentShardID, currentKeys, heartbeatDetails)
+		if err != nil {
+			return nil, err
+		}
+		heartbeatDetails = FixShardHeartbeatDetails{
+			LastShardIndexHandled: i,
+			Reports:               append(heartbeatDetails.Reports, *shardReport),
+		}
+	}
+	return heartbeatDetails.Reports, nil
+}
+
+func fixShard(
+	activityCtx context.Context,
+	params FixShardActivityParams,
+	shardID int,
+	corruptedKeys store.Keys,
+	heartbeatDetails FixShardHeartbeatDetails,
+) (*FixReport, error) {
+	ctx := activityCtx.Value(params.ContextKey).(FixerContext)
+	resources := ctx.Resource
+	scope := ctx.Scope.Tagged(metrics.ActivityTypeTag(ActivityFixShard))
+	sw := scope.StartTimer(metrics.CadenceLatency)
+	defer sw.Stop()
+
+	if ctx.Hooks == nil {
+		return nil, errors.New("fixer hooks are not provided")
```

Comment:
I think the fixShardActivity and scanShardActivity has a retry policy on them. If that is the case then I think we want to make sure we are treating this case as a non-retryable error.

Reply:
good catch. Added non-retryable error.